Unsafe mechanism for turning a primitive into an enum?

I have a situation where I know I have a numbered `enum` where I know that I have a valid primitive for that enum and I'd rather avoid the cost of the `try_into`. Is it plausible/reasonable to add an `unsafe` method for converting primitives into an `enum`? Maybe `UnsafeFromPrimitive`?

> `transmute` [...] almost entirely bypasses the type system (e.g. it will merrily let me transmute a `usize` to a `u8`).

That's not the case; `transmute` will error if the sizes of types aren't the same:
```
error[E0512]: cannot transmute between types of different sizes, or dependently-sized types
 --> src/main.rs:3:9
  |
3 |         std::mem::transmute::<usize, u8>(1_usize);
  |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: source type: `usize` (64 bits)
  = note: target type: `u8` (8 bits)

error: aborting due to previous error

For more information about this error, try `rustc --explain E0512`.
error: Could not compile `rustplay`.

To learn more, run the command again with --verbose.
```

so you can probably transmute a `u64` and a `usize`, but you can't pass around arbitrary types.

It sounds like you want explicit, rather than inferred, types on the `transmute` call, and that the boilerplate you're looking to save is some combination of writing the turbofish, and having to update the type in the turbofish if the `repr` of the `enum` changes. I think I buy that that's worthwhile.

How does #7 look to you?

--- a/README.md
+++ b/README.md
@@ -54,6 +54,39 @@ fn convert() {
 }
 ```
 
+Unsafely turning a primitive into an enum with from_unchecked
+-------------------------------------------------------------
+
+If you're really certain a conversion will succeed, and want to avoid a small amount of overhead, you can use unsafe
+code to do this conversion. Unless you have data showing that the match statement generated in the `try_from` above is a
+bottleneck for you, you should avoid doing this, as the unsafe code has potential to cause serious memory issues in
+your programme.
+
+```rust
+use num_enum::UnsafeFromPrimitive;
+
+#[derive(Debug, Eq, PartialEq, UnsafeFromPrimitive)]
+#[repr(u8)]
+enum Number {
+    Zero,
+    One,
+}
+
+#[test]
+fn convert() {
+    unsafe {
+        assert_eq!(
+            Number::Zero,
+            Number::from_unchecked(0_u8)
+        );
+        assert_eq!(
+            Number::One,
+            Number::from_unchecked(1_u8)
+        );
+    }
+}
+```
+
 Optional features
 -----------------
 

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -81,7 +81,7 @@ pub fn derive_unsafe_from_primitive(stream: TokenStream) -> TokenStream {
 
     let expanded = quote! {
         impl #name {
-            pub unsafe fn from(num: #repr) -> Self {
+            pub unsafe fn from_unchecked(num: #repr) -> Self {
                 ::core::intrinsics::transmute(num)
             }
         }

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -74,6 +74,22 @@ pub fn derive_try_from_primitive(stream: TokenStream) -> TokenStream {
     TokenStream::from(expanded)
 }
 
+#[proc_macro_derive(UnsafeFromPrimitive)]
+pub fn derive_unsafe_from_primitive(stream: TokenStream) -> TokenStream {
+    let input = parse_macro_input!(stream as DeriveInput);
+    let EnumInfo { name, repr, .. } = parse_enum(input);
+
+    let expanded = quote! {
+        impl #name {
+            pub unsafe fn from(num: #repr) -> Self {
+                ::core::intrinsics::transmute(num)
+            }
+        }
+    };
+
+    TokenStream::from(expanded)
+}
+
 struct EnumInfo {
     name: syn::Ident,
     repr: proc_macro2::Ident,

--- a/tests/lib.rs
+++ b/tests/lib.rs
@@ -274,3 +274,26 @@ fn error_variant_is_allowed() {
         Err("No value in enum HasErrorVariant for value 2".to_owned())
     );
 }
+
+use num_enum::UnsafeFromPrimitive;
+
+#[derive(Debug, Eq, PartialEq, UnsafeFromPrimitive)]
+#[repr(u8)]
+enum HasUnsafeFromPrimitiveNumber {
+    Zero,
+    One,
+}
+
+#[test]
+fn has_unsafe_from_primitive_number() {
+    unsafe {
+        assert_eq!(
+            HasUnsafeFromPrimitiveNumber::Zero,
+            HasUnsafeFromPrimitiveNumber::from(0_u8)
+        );
+        assert_eq!(
+            HasUnsafeFromPrimitiveNumber::One,
+            HasUnsafeFromPrimitiveNumber::from(1_u8)
+        );
+    }
+}

--- a/tests/lib.rs
+++ b/tests/lib.rs
@@ -289,11 +289,11 @@ fn has_unsafe_from_primitive_number() {
     unsafe {
         assert_eq!(
             HasUnsafeFromPrimitiveNumber::Zero,
-            HasUnsafeFromPrimitiveNumber::from(0_u8)
+            HasUnsafeFromPrimitiveNumber::from_unchecked(0_u8)
         );
         assert_eq!(
             HasUnsafeFromPrimitiveNumber::One,
-            HasUnsafeFromPrimitiveNumber::from(1_u8)
+            HasUnsafeFromPrimitiveNumber::from_unchecked(1_u8)
         );
     }
 }

